Process Category Pages

ebenezer-melkamu opened the pull request:
Fixes #272 

### Modifications 
1. Added a `parser_utils` module and isolated util modules from `parser.py` to the new module


* [ ] the country page (britain.dita) will need to be modified to create the correct XREF urls
* [ ] fixes as commented in this commit: https://github.com/DeepBlueCLtd/LegacyMan/commit/b86cc0111be9e56a21d4efa2d847ebbc6b2bb315

ebenezer-melkamu committed:
Add util functions to parser_utils module

IanMayo commented:
> I'd like us to try this strategy:

Note for @AbegaM .  "if" this alternative strategy works, I think it could sit in the `utils` module, since it's not specific to `LegacyMan` - it will be a general purpose `html` to `dita` converter function.

Note: I did look for an existing html-to-dita converter, but couldn't find one.

ebenezer-melkamu commented:
@IanMayo There are this `<colspec>` elements in the `.dita` file, should i generate these elements manually using a loop? i don't see this `<colspec>` elements reflected in the `html` file we are parsing

[colespec element](https://github.com/DeepBlueCLtd/Fi3ldMan/blob/4fe9374280c1c40afabb5c9029ef14be10485cf2/dita/regions/britain/legacy/britain_legacy.dita#L13)

IanMayo commented:
I'm checking if they are essential.

IanMayo commented:
For now, let's assume the are compulsory.  For this `classlist`, the table will always be 7 columns wide.  So, you could hard-code these colspecs and ids.

IanMayo commented:
Aah, we do need the specific names, in order for the heading row to occupy all columns:
```
 <entry namest="col1" nameend="col7">Overview of systems in France</entry>
```

ebenezer-melkamu commented:
@IanMayo after creating the `.dita` files where should we store them, for example i have this `.html` files
  ```
     ../Britain_Complex/Britain_Complex.html
     ../Britain_Standalones/Britain_Standalones.html
     ../Britain_Composite/Britain_Composites.html
     ../Britain_Legacy/Britain_Legacy.html
  ```
Should we store them in this kind of path 
  ```
     dita/regions/britain/britain_complex/britain_complex.dita
     dita/regions/britain/britain_standalones/britain_standalones.dita
  ```

IanMayo commented:
Yes that's right.

ebenezer-melkamu committed:
Generate category pages + Add function to generate category pages

ebenezer-melkamu commented:
@IanMayo I have generated the category pages, i do need to refine my code, but please run the generated code on your end and let me know what you see

IanMayo commented on parser/parser.py:
```diff
@@ -173,6 +177,97 @@ def process_ns_countries(country_name, link):
+       raise ValueError(f"<td> element with colspan 7 not found in the {category_page_link} file")
+
+    #Create the DITA document type declaration string
+    dita_doctype = '<!DOCTYPE classlist SYSTEM "../../../dtd/classlist.dtd">'
```
note: you need a two more '../` levels to get to `classlist.dtd`.

IanMayo commented on target/dita/regions/Iceland/iceland_complex/iceland_complex.dita:
```diff
@@ -0,0 +1,56 @@
+	<title>Iceland - Complex</title>
+	<flag>
+		<fig>
+			<image alt="flag" href="images/iceland_flag.png"/>
```
this line needs to be `../content/images/Flag.png`

ebenezer-melkamu replied:
ok, but when i check the folder `britain/content/images` i see a `Flag.jpg` file there but not `Flag.png` and when i check the `iceland/content/images` folder i do have a `Flag.png` file

IanMayo replied:
I see...  Investigating.

Ok here is the logic to find the URL of the flag:
https://github.com/DeepBlueCLtd/LegacyMan/blob/develop/legacyman_parser/parse_countries.py#L131

So, in the NS-Country page (britain.html), find the `h2`. Then find the next `img` file.  That has the URL of the flag.  Pass that URL as a parameter to the new method, and use that.

IanMayo committed:
fixes

ebenezer-melkamu committed:
Modify process_category_pages function to add id, namest, nameend attr + Change href for flag picture + Regenerate legacy data

ebenezer-melkamu committed:
Fix merge conflict

ebenezer-melkamu commented:
Hello @IanMayo, i have pushed a new change, i have fixed the code based on the comments, please run the code from your end and let me know if there is anything i have to fix

IanMayo commented:
It looks really good. Should be at my desk in 30 mins to double-check.

You should start thinking about the next one, #274  :-D

IanMayo commented:
One fix. The country pages need to have the hrefs updated, so they point to the new category page, rather than `self`.

Aah, I _expect_ we can now use the correct image too.  The URLs for the image should be in the `./[country-category]/content/images` folder.

So, we will replace the instances of `./britain.dita` and the `world-map.gif` in this snippet:
![image](https://github.com/DeepBlueCLtd/LegacyMan/assets/1108513/47bdd58a-fca1-44a3-acde-f9cce9caeaaf)

ebenezer-melkamu committed:
Add util functions + Modify link for xref and image

ebenezer-melkamu commented:
@IanMayo I was having errors when i changed some links, it turns out some of the file names for the images in the `/content/images` folder were uppercase like `FlagS.jpg` but now i have changed everything to lower case and i have also fixed the code based on the above comment

IanMayo commented:
Hello, just two issues

* [ ] In the category-list page, where the is a link - we don't need the duplicate text label. See `unit_daf` below:
![image](https://github.com/DeepBlueCLtd/LegacyMan/assets/1108513/3e4f9683-ee60-4531-be16-088888d89311)
* [ ] There can be `a` elements that are anchors, not links.   We only process it as a link if it has an `href`. We don't create a link for a file like [this](https://github.com/DeepBlueCLtd/LegacyMan/blob/develop/data/Britain_Standalones/Britain_Standalones.html#L30).  When we encounter an `a` anchor, we should create a DITA `<anchor>` like [this](https://docs.oasis-open.org/dita/dita/v1.3/os/part1-base/langRef/base/anchor.html) - using `name` as `id`.

ebenezer-melkamu committed:
Remove duplicate text from <entry>

IanMayo commented:
Nice work @AbegaM - now just the `<anchors>`.

ebenezer-melkamu commented:
@IanMayo I am on it, i will push another change in a minute 👍

ebenezer-melkamu commented:
@IanMayo I am getting an error saying `"anchor must be declared"`, i think i i have to download a `.dtd` file for the `anchor` element right?

IanMayo commented:
> @IanMayo I am getting an error saying `"anchor must be declared"`, i think i i have to download a `.dtd` file for the `anchor` element right?

No, let's kick that can down the road :-D   Could you create an issue here in LegacyMan for it please?

I'll merge this issue.
